hail/src/main/scala/is/hail/cxx/Emit.scala

```diff
@@ -857,6 +856,56 @@ class Emitter(fb: FunctionBuilder, nSpecialArgs: Int, ctx: SparkFunctionContext)
              |})
              |""".stripMargin)
 
+      case ir.NDArrayMap(child, elemName, body) =>
+        val elemPType = child.pType.asInstanceOf[PNDArray].elementType
+        val cxxElemType = typeToCXXType(elemPType)
+        val newElemPType = body.pType
+        val newDataContainer = PArray(newElemPType)
+
+        val ndt = emit(child)
+        val nd = fb.variable("nd", "NDArray", ndt.v)
+
+        val elemRef = fb.variable("elemRef", cxxElemType)
+        val bodyt = outer.emit(body,
+          env.bind(elemName, EmitTriplet(elemPType, "", "false", elemRef.toString, resultRegion)))
+
+        val length = fb.variable("length", "long", s"n_elements($nd.shape)")
+        val sab = resultRegion.arrayBuilder(fb, newDataContainer)
+        val data = fb.variable("data", "const char *")
+
+        val childPretty = StringEscapeUtils.escapeString(ir.Pretty.short(child))
+        val bodyPretty = StringEscapeUtils.escapeString(ir.Pretty.short(body))
+        present(
+          s"""
+             |({
+             | ${ nd.define }
+             | ${ length.define }
+             | ${ elemRef.define }
+             | ${ data.define }
+             |
+             | ${ ndt.setup }
+             | if (${ ndt.m }) {
+             |   ${ fb.nativeError("NDArray cannot be missing. IR: %s".format(childPretty)) }
+             | }
```

Comment:
Moved the define. Are you not describing the point of `present`? Why would missing always be `true`? I can remove the check for `ndt` since it is an invariant of NDArrays right now.

Reply:
oops, s/`true`/`false`/.

Yeah, basically the point of present---and if you say that all NDArray emit triplets need to be defined in that format, you can just skip the calls to `setup` and `m` for any NDArray emit triplet.